### ISSUE TITLE
Better estimate of the solution of a non linear solid mechanic problem using a tangent linear resolution

### ISSUE THREAD

**Issue by thelfer** [mfem/mfem#2174]
Dear all,

I solve mechanical non linear problems with mfem using the default Newton solver. 

My trouble is that the imposition of the Dirichlet boundary conditions leads to an initial inappropriate estimate of the solution with a very high strain estimate on the boundary elements.

### Description of the issue

To illustrate the problem, consider a beam in tension with two elements (1 and 2) and two boundaries L and R:

    --------------
 L |    1   |  2  | R
    --------------

The normal displacement on L is 0. The normal displacement on R is imposed to U.

The Newton solver uses as an initial estimate of the solution a displacement field where the only the nodes on boundary R are moved by the displacement U. Hence, the Element 1 is undeformed and Element 2 may be highly deformed. This may lead to divergence of the mechanical behaviour or divergence of the Newton method on real structures (in my case, a notched beam in finite strain elasto-plasticity).

This obvisouly gets  worse a one refines the mesh as the strain in the last element is given by U/h where h is size of the last element.

### A solution ?

A somehow classical solution would be to solve the so-called tangent problem to predict a more physical estimate. For a given time step, the idea would be to assemble the tangent matrix K and then solve a linear problem using this matrix by imposing only the increment of the imposed displacement on boundaries affected by Dirichlet boundary conditions. In this auxiliary problem, the mechanical behaviour integration is not called and the increment of imposed displacement are smeared over the whole structure. The standard Newton solver would then be called.

### Implementation issue

In pratice, I have:

- The solution u of the previous time step
- A non linear form `F`.
- A Linear solver `L`
- The list `ess_dofs_lits` of dofs affected by the Dirichlet boundary conditions
- An vector `x` which is zero everywhere exept where  the value of the increment of the displacement is known.

I can call the `GetGradient` method of `F` to get an `Operator` K.
~~~~
auto& K = F.GetGradient(u);
~~~~

And then I am stuck. I have two issues here:

- The Dirichlet boundary conditions have already been handled by the `GetGradient` method which seems to be problematic in my case. A simple workaround would be to remove the Dirichlet boundary conditions before calling  `GetGradient` and set them back after. Is it what I shall do ?
- I don't know how to form a linear system from my `Operator` K to solve my linear problem. This seems a trivial question, but I don't see the solution.

Thanks for any help,

Regards,

Thomas Helfer

**Comment by jandrej** [comment]
Is it feasible to convert your boundary condition into a time dependent ramp?

ping @rcarson3 @jamiebramwell

**Comment by thelfer** [comment]
Hi @jandrej 

What I do currently is that I treat divergences of the Newton solver by substepping, i.e. applying small increments of the displacement per time step (time has no physical meaning here, because I treat a plastic behaviour). And it works.

However, this leads to much smaller time steps compared to other solvers in which I also treat this case, so it is mostly a performance issue.

Thomas

**Comment by rcarson3** [comment]
@thelfer so I believe I understand what you're talking about. I solved this by doing what I listed in this comment:

https://github.com/LLNL/ExaConstit/issues/8#issuecomment-708074407
which is similar to what @v-dobrev outlined in a comment here:
https://github.com/mfem/mfem/issues/1720#issuecomment-709505317
The PR that goes over my implementation is found here which might help you:
https://github.com/LLNL/ExaConstit/pull/26

I'm pretty much doing what you're talking about in your original comment, but it does require you to do a bit more work on your end of things rather than within MFEM in order to get things to work. I found that I had to have a step that applied the guess first to get the solution going in the right direction and afterwards I could just call my regular NR set of methods.

**Comment by thelfer** [comment]
@rcarson3 Thanks again for your help ! If we happen to meet in a conference some day, I definitively owe you a beer (or anything you prefer). I'll definitively study your links this evening.

**Comment by thelfer** [comment]
However, just to be able to play around, I really would like to be able to have some hints for my issues:

> I have two issues here:
>
> The Dirichlet boundary conditions have already been handled by the GetGradient method
> which seems to be problematic in my case. A simple workaround would be to remove the
> Dirichlet boundary conditions before calling GetGradient and set them back after. Is it what I shall do ?

Is there any issue with the idea of setting/unsetting essential boundary conditions on a non linear form ?

> I don't know how to form a linear system from my Operator K to solve my linear problem.
> This seems a trivial question, but I don't see the solution.

Any idea for this *a priori* simple question ?

**Comment by rcarson3** [comment]
@thelfer so in my changing BC pr I linked to I change the boundary conditions/essential boundary attributes every step without it affecting the solution. I believe @v-dobrev has suggested to me before in the past that setting and unsetting the attributes would be the easiest way to get the full gradient matrix out of the nonlinear form to be used as described in one of my earlier links.

>>I don't know how to form a linear system from my Operator K to solve my linear problem.
This seems a trivial question, but I don't see the solution.

I'm not quite too sure I follow what your talking about here. Do you mean getting out the matrix/gradient term supplied by the NonlinearForm::GetGradient call or are you talking about something else? If your talking about `NonlinearForm::GetGradient` after unsetting the essential boundary attributes you can just call this function as you would normally and get the operator and do with it what as you need to do.

**Comment by thelfer** [comment]
> I'm not quite too sure I follow what your talking about here.

So it's probably that I am not clear enough.

> If your talking about `NonlinearForm::GetGradient` after unsetting the essential boundary attributes you can just call this function as you would normally and get the operator and do with it what as you need to do.

My question is basically the following. I have `K` as a result from get GetGradient method without the boundary conditions, the list of boundary conditions `ess_dofs_list`, a vector `x` which is zero everywhere except on the dofs where an increment of displacement is prescribed and `L` a linear solver. I jus want to compute the solution of `K^{-1}\,x` taking the boundary conditions in to account using `L`...

Probably, a very basic question, but I don't know how to do it !

**Comment by rcarson3** [comment]
Okay so I think you want to do something like what's done here:

The first function the whole entire initial guess when you have changing boundary conditions:
https://github.com/LLNL/ExaConstit/blob/exaconstit-dev/src/system_driver.cpp#L178-L204
and this one gets out the necessary gradient matrix and modifies the LHS vectors to have the necessary values that they should:
https://github.com/LLNL/ExaConstit/blob/exaconstit-dev/src/mechanics_operator.cpp#L461-L511

I will note that `Hform->GetLocalGradient2(x)` listed above just set-ups the K matrix to be what it should be without the boundary conditions, since I didn't go down the remove the essential boundary attribute list for mine.

I hope this helps a bit more.

**Comment by v-dobrev** [comment]
Is it not possible to define analytically a **smooth** initial deformation that satisfies the displacement boundary conditions? Maybe simply interpolate linearly between the left boundary and the right boundary. Using such deformation as the initial guess for the Newton solve maybe good enough. The initial deformation can be initialized similar to here (with properly modified `InitialDeformation` function): https://github.com/mfem/mfem/blob/f4a99fa854c55554c1a48879077ef34ff69f6188/examples/ex10p.cpp#L324-L325

**Comment by thelfer** [comment]
@v-dobrev This is precisely the idea of the tangent prediction :)
I am currently studying the code proposed by @rcarson3.
But I also a bit stubborn and wanted to achieve my ideas by myself but still haven't not found a solution to my problem. I am a bit frustrated by the fact that I can now do rather complex and highly non linear computations within my application but can't figure out how to solve a simple linear problem. 

See this question:

> My question is basically the following. I have K as a result from get GetGradient method without the boundary conditions,
> the list of boundary conditions ess_dofs_list, a vector x which is zero everywhere except on the dofs where an increment of
> displacement is prescribed and L a linear solver. I jus want to compute the solution of this linear problem taking the
> boundary conditions in to account using L...

K being an object of type `Operator`, is there an example of how to eliminate the columns and the rows associated with essential boundary conditions, update the rhs and just call the linear solver ?

**Comment by v-dobrev** [comment]
To impose b.c. at the `Operator` level, one can use the matrix-free versions of `FormLinearSystem`: https://github.com/mfem/mfem/blob/96fc6f177ea8f2174676e7dc6d5c9ec1a08a4d60/linalg/operator.hpp#L167-L170
and `RecoverFEMSolution`: https://github.com/mfem/mfem/blob/96fc6f177ea8f2174676e7dc6d5c9ec1a08a4d60/linalg/operator.hpp#L212

However, your choice of solver/preconditioner is limited in this case. It is probably better to figure out the specific format used by `K` -- I suspect it is a `HypreParMatrix` if it was obtained from `ParNonlinearForm::GetGradient`. Then you can use methods like https://github.com/mfem/mfem/blob/96fc6f177ea8f2174676e7dc6d5c9ec1a08a4d60/linalg/hypre.hpp#L567-L570 followed by https://github.com/mfem/mfem/blob/96fc6f177ea8f2174676e7dc6d5c9ec1a08a4d60/linalg/hypre.hpp#L645-L649 to perform what is essentially `FormLinearSystem` for a `HypreParMatrix`. Note that the `HypreParMatrix` is a true-dof size `Operator` (i.e. it acts on T-vectors, see https://mfem.org/performance/#finite-element-operator-decomposition) and you probably need to perform prolongations/restrictions to transform the solution/rhs if those are given as L-vectors like `ParGridFunction` or `ParLinearForm`. You can take a look here: https://github.com/mfem/mfem/blob/96fc6f177ea8f2174676e7dc6d5c9ec1a08a4d60/fem/pbilinearform.cpp#L340-L342 and here https://github.com/mfem/mfem/blob/96fc6f177ea8f2174676e7dc6d5c9ec1a08a4d60/fem/pbilinearform.cpp#L444-L445 to see how this is done in `ParBilinearForm`.

This approach for obtaining an initial guess for Newton is not conveniently supported by MFEM at the moment but it may be worth adding. What do you think @thelfer, @jandrej, @rcarson3?

**Comment by thelfer** [comment]
@v-dobrev Thanks for the tips. Finally, it seems like that it was not a so trivial question after all. Do you think that an helper function such as `SolveLinearSystem` would be of any help for `MFEM` users ?

Something like:

```cxx
/*!
 * \brief solve A x = b taking essential boundary conditions into account
 * \param[in,out] x
 * \param[in] A: 
 * \param[in] l: linear solver
 * \param[in] ess_dofs_list:
 * \param[in] b:
 * \return  the exit status of the linear solver (see https://github.com/mfem/mfem/issues/2139)
 */
int SolveLinearSystem(Vector&, Operator&, Solver&, Array<int>&, const Vector&);
```

> This approach for obtaining an initial guess for Newton is not conveniently supported by MFEM 
> at the moment but it may be worth adding. What do you think @thelfer, @jandrej, @rcarson3?

It definitively can be useful, but other strategies can be more appropriate depending on the problem treated. For example, an elastic prediction is sometimes more appropriate than a tangent prediction

For the discussion, what about enhancing the NewtonSolver interface by:

1. Defining some constant defining the default prediction policies:
```cxx
static constexpr int NOPREDICTION = 0;
static constexpr int TANGENTPREDICTION = 1;
```
2. Adding a data member `prediction_policy` initialized to `NOPREDICTION` and a member function `SetPredictionPolicy`
3. Adding  two methods called respectively `ComputePrediction` and `ComputeTangentPrediction`. The `ComputePrediction` 
could be called at the very beginning of the `Mult` method and would:
    - Do nothing if `prediction_policy` is set to `NOPREDICTION`
    - Call a method called `ComputeTangentPrediction` if `prediction_policy` is set to `TANGENTPREDICTION`.
    - Fail otherwise

That design would let the user override the `ComputePrediction` method to implement other prediction policies.

Of course, this is just live thinking, so do not hesitate to critisize it. @jandrej, @rcarson3, any comment ?

**Comment by v-dobrev** [comment]
I'm not sure how exactly this predictions will work. In order to compute `ParNonlinearForm::GetGradient(x)` you still need a reasonable state `x`. Where do you get that? If that `x` is reasonable for `GetGradient`, isn't it also reasonable for the Newton initial guess? Maybe the `x` for `GetGradient` does not need to satisfy the essential b.c. -- is that the difference?

**Comment by thelfer** [comment]
@v-dobrev You perfectly have understood the point. I mostly work with evolution problem, so the solution of the previous time step is generally a good estimate. One may also try to predict the solution by extrapolating solutions of previous time steps, but it generally on works on monolitcally increasing radial loadings, i.e. the loading increases but the direction of the loading does not change too much.

 In my case, `x` isn't even needed because I store the state of the material at the begining of the  time step, including the strain and/or deformation gradient, which is enough to compute the tangent operator. 

In the hyperelastic case, `x` would be the displacement at the beginning of the time step, not taking into account essential b.c., which is enough to build the tangent operator.

In this perspective, it is clear that the evolution of Neuman-like boundary conditions (so called external forces, like imposed pressure) is not considered here, which is debatable. IMHO, those boundary conditions does not lead to such terrible difficulties that I highlighted for essential b.c., so discarding them by default is quite fair: people willing to take them into account could override the `ComputeTangentPrediction` method anyway.

**Comment by rcarson3** [comment]
> For the discussion, what about enhancing the NewtonSolver interface by:
> 
>     1. Defining some constant defining the default prediction policies:
> 
> 
> ```c++
> static constexpr int NOPREDICTION = 0;
> static constexpr int TANGENTPREDICTION = 1;
> ```
> 
>     1. Adding a data member `prediction_policy` initialized to `NOPREDICTION` and a member function `SetPredictionPolicy`
> 
>     2. Adding  two methods called respectively `ComputePrediction` and `ComputeTangentPrediction`. The `ComputePrediction`
>        could be called at the very beginning of the `Mult` method and would:
>        
>        * Do nothing if `prediction_policy` is set to `NOPREDICTION`
>        * Call a method called `ComputeTangentPrediction` if `prediction_policy` is set to `TANGENTPREDICTION`.
>        * Fail otherwise
> 
> 
> That design would let the user override the `ComputePrediction` method to implement other prediction policies.

So, I would rather see the `NewtonSolver` kept the same as it is just because it's a fairly basic class that users can make use of. I could see it useful to potentially have a more advance example or potentially a small miniapp that covers some of the issues being discussed here as some of them can be a bit more domain specific. Now if there ends up being enough users that find these features useful then I think it might be beneficial to introduce a derived `NewtonSolver` class into the main library that contains those features. 

Although, I do think specific features such as a pleasant way to obtain the gradient operator without the boundary conditions being applied would be nice additions to `NonlinearForm` and related classes. I've got that code lying around in my fork of MFEM, so it shouldn't be hard. However, I'm a bit busy at the moment with other projects, so I would need to find the time to do so :/

**Comment by thelfer** [comment]
@chunyu79 Thanks for your indications.

I finally took the time to work on this issue in `MFEM/MGIS`. I basically choose solve a linear problem based on the gradient at the beginning of the time time step to find an estimate of the displacement increment (in mechanicals). The tricky part was to create linear forms for nonlinear forms.

I did not reuse your (elegant) solution because I wanted to be able to:

1. update the external forces (imposed pressure for instance)
2. take evolution of stress-free strain into account (thermal expansion for instance)
3. have more fexibility in the choice of the prediction operator (use the elastic operator for instance).
